### PR TITLE
Rewrote SuggestsCommand

### PR DESCRIPTION
**Pull request by curry684**
I rewrote the `SuggestsCommand.php` as it had multiple flaws:

- Default output wasn't sorted by anything visible, making it nearly unusable in projects of any size
- It did not take `provide` and `replace` sections into account, so for example in a Symfony project it would keep suggesting `symfony/console` even though that's provided by `symfony/symfony` through a `replace`.
- There were some potential exponential performance issues because of nested loops on the list of installed packages - new implementation should be a ton faster on large projects

New version introduces 2 new flags:

      --by-package               Groups output by suggesting package
      --by-suggestion            Groups output by suggested package

These show a grouped minitree of the suggested packages, grouped both ways. Default is to just dump a (now sorted) flat list of suggested packages. `-v` now implies all possible options (both reports).

### CODE REVIEW

**Comment by curry684** [comment]
Sample output when using this root require:

    "require": {
        "symfony/symfony": "^3.0",
        "jms/serializer": "^1.1",
        "monolog/monolog": "^1.17",
        "kriswallsmith/assetic": "^1.3"
    }

Outputs:

    C:\Projects\ComposerTest>php ..\Composer\bin\composer suggests
    aws/aws-sdk-php
    doctrine/couchdb
    ext-amqp
    ext-libsodium
    ext-mbstring
    ext-mongo
    graylog2/gelf-php
    leafo/lessphp
    leafo/scssphp
    leafo/scssphp-compass
    patchwork/jsqueeze
    php-console/php-console
    ptachoire/cssembed
    raven/raven
    rollbar/rollbar
    ruflin/elastica
    videlalvaro/php-amqplib



    C:\Projects\ComposerTest>php ..\Composer\bin\composer suggests --by-package
    kriswallsmith/assetic suggests:
     - leafo/lessphp: Assetic provides the integration with the lessphp LESS compiler
     - leafo/scssphp: Assetic provides the integration with the scssphp SCSS compiler
     - leafo/scssphp-compass: Assetic provides the integration with the SCSS compass plugin
     - patchwork/jsqueeze: Assetic provides the integration with the JSqueeze JavaScript compressor
     - ptachoire/cssembed: Assetic provides the integration with phpcssembed to embed data uris

    monolog/monolog suggests:
     - aws/aws-sdk-php: Allow sending log messages to AWS services like DynamoDB
     - doctrine/couchdb: Allow sending log messages to a CouchDB server
     - ext-amqp: Allow sending log messages to an AMQP server (1.0+ required)
     - ext-mongo: Allow sending log messages to a MongoDB server
     - graylog2/gelf-php: Allow sending log messages to a GrayLog2 server
     - php-console/php-console: Allow sending log messages to Google Chrome
     - raven/raven: Allow sending log messages to a Sentry server
     - rollbar/rollbar: Allow sending log messages to Rollbar
     - ruflin/elastica: Allow sending log messages to an Elastic Search server
     - videlalvaro/php-amqplib: Allow sending log messages to an AMQP server using php-amqplib

    paragonie/random_compat suggests:
     - ext-libsodium: Provides a modern crypto API that can be used to generate random bytes.

    symfony/polyfill-mbstring suggests:
     - ext-mbstring: For best performance



    C:\Projects\ComposerTest>php ..\Composer\bin\composer suggests --by-suggestion
    aws/aws-sdk-php is suggested by:
     - monolog/monolog: Allow sending log messages to AWS services like DynamoDB

    doctrine/couchdb is suggested by:
     - monolog/monolog: Allow sending log messages to a CouchDB server

    ext-amqp is suggested by:
     - monolog/monolog: Allow sending log messages to an AMQP server (1.0+ required)

    ext-libsodium is suggested by:
     - paragonie/random_compat: Provides a modern crypto API that can be used to generate random bytes.

    ext-mbstring is suggested by:
     - symfony/polyfill-mbstring: For best performance

    ext-mongo is suggested by:
     - monolog/monolog: Allow sending log messages to a MongoDB server

    graylog2/gelf-php is suggested by:
     - monolog/monolog: Allow sending log messages to a GrayLog2 server

    leafo/lessphp is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with the lessphp LESS compiler

    leafo/scssphp is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with the scssphp SCSS compiler

    leafo/scssphp-compass is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with the SCSS compass plugin

    patchwork/jsqueeze is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with the JSqueeze JavaScript compressor

    php-console/php-console is suggested by:
     - monolog/monolog: Allow sending log messages to Google Chrome

    ptachoire/cssembed is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with phpcssembed to embed data uris

    raven/raven is suggested by:
     - monolog/monolog: Allow sending log messages to a Sentry server

    rollbar/rollbar is suggested by:
     - monolog/monolog: Allow sending log messages to Rollbar

    ruflin/elastica is suggested by:
     - monolog/monolog: Allow sending log messages to an Elastic Search server

    videlalvaro/php-amqplib is suggested by:
     - monolog/monolog: Allow sending log messages to an AMQP server using php-amqplib


    C:\Projects\ComposerTest>php ..\Composer\bin\composer suggests -v
    kriswallsmith/assetic suggests:
     - leafo/lessphp: Assetic provides the integration with the lessphp LESS compiler
     - leafo/scssphp: Assetic provides the integration with the scssphp SCSS compiler
     - leafo/scssphp-compass: Assetic provides the integration with the SCSS compass plugin
     - patchwork/jsqueeze: Assetic provides the integration with the JSqueeze JavaScript compressor
     - ptachoire/cssembed: Assetic provides the integration with phpcssembed to embed data uris

    monolog/monolog suggests:
     - aws/aws-sdk-php: Allow sending log messages to AWS services like DynamoDB
     - doctrine/couchdb: Allow sending log messages to a CouchDB server
     - ext-amqp: Allow sending log messages to an AMQP server (1.0+ required)
     - ext-mongo: Allow sending log messages to a MongoDB server
     - graylog2/gelf-php: Allow sending log messages to a GrayLog2 server
     - php-console/php-console: Allow sending log messages to Google Chrome
     - raven/raven: Allow sending log messages to a Sentry server
     - rollbar/rollbar: Allow sending log messages to Rollbar
     - ruflin/elastica: Allow sending log messages to an Elastic Search server
     - videlalvaro/php-amqplib: Allow sending log messages to an AMQP server using php-amqplib

    paragonie/random_compat suggests:
     - ext-libsodium: Provides a modern crypto API that can be used to generate random bytes.

    symfony/polyfill-mbstring suggests:
     - ext-mbstring: For best performance

    ------------------------------------------------------------------------------
    aws/aws-sdk-php is suggested by:
     - monolog/monolog: Allow sending log messages to AWS services like DynamoDB

    doctrine/couchdb is suggested by:
     - monolog/monolog: Allow sending log messages to a CouchDB server

    ext-amqp is suggested by:
     - monolog/monolog: Allow sending log messages to an AMQP server (1.0+ required)

    ext-libsodium is suggested by:
     - paragonie/random_compat: Provides a modern crypto API that can be used to generate random bytes.

    ext-mbstring is suggested by:
     - symfony/polyfill-mbstring: For best performance

    ext-mongo is suggested by:
     - monolog/monolog: Allow sending log messages to a MongoDB server

    graylog2/gelf-php is suggested by:
     - monolog/monolog: Allow sending log messages to a GrayLog2 server

    leafo/lessphp is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with the lessphp LESS compiler

    leafo/scssphp is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with the scssphp SCSS compiler

    leafo/scssphp-compass is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with the SCSS compass plugin

    patchwork/jsqueeze is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with the JSqueeze JavaScript compressor

    php-console/php-console is suggested by:
     - monolog/monolog: Allow sending log messages to Google Chrome

    ptachoire/cssembed is suggested by:
     - kriswallsmith/assetic: Assetic provides the integration with phpcssembed to embed data uris

    raven/raven is suggested by:
     - monolog/monolog: Allow sending log messages to a Sentry server

    rollbar/rollbar is suggested by:
     - monolog/monolog: Allow sending log messages to Rollbar

    ruflin/elastica is suggested by:
     - monolog/monolog: Allow sending log messages to an Elastic Search server

    videlalvaro/php-amqplib is suggested by:
     - monolog/monolog: Allow sending log messages to an AMQP server using php-amqplib

File: src/Composer/Command/SuggestsCommand.php [hunk 1]
```diff
@@ -55,44 +57,85 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+            $mode |= 2;
+        }
+        if ($io->isVerbose()) {
+            $mode = ~0;
+        }
 
-            if (empty($reason)) {
-                $reason = '*';
+        // Simple mode
+        if ($mode == 0) {
```
Comment (stloyd) [inline src/Composer/Command/SuggestsCommand.php]: Should be strict & yoda:
```php
if (0 === $mode) {}

File: src/Composer/Command/SuggestsCommand.php [hunk 2]
```diff
@@ -55,44 +57,85 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
         $filter = $input->getArgument('packages');
 
-        foreach ($packages as $package) {
-            if (empty($package['suggest'])) {
-                continue;
-            }
+        // First assemble list of packages that are installed, replaced or provided
+        $installed = array();
+        foreach($packages as $package) {
+            $installed[] = $package['name'];
```
Comment (stloyd) [inline src/Composer/Command/SuggestsCommand.php]: To remove usage of `in_array()` below you could use names as array keys & later just simply `isset()`.

File: src/Composer/Command/SuggestsCommand.php [hunk 3]
```diff
@@ -55,44 +57,85 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
-    {
-        foreach ($suggestions as $suggestion => $reason) {
-            foreach ($installed as $package) {
-                if ($package['name'] === $suggestion) {
-                    continue 2;
+        // Next gather all suggestions that are not in that list
+        $suggesters = array();
+        $suggested = array();
+        foreach ($packages as $package) {
+            if ((empty($filter) || in_array($package['name'], $filter)) && !empty($package['suggest'])) {
```
Comment (stof) [inline src/Composer/Command/SuggestsCommand.php]: I would rather use the opposite condition and ``continue``, which would reduce the indentation

File: src/Composer/Command/SuggestsCommand.php [hunk 4]
```diff
@@ -55,44 +57,89 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        // Undub and sort the install list into a sorted lookup array
+        $installed = array_flip($installed);
+        ksort($installed);
+
+        // Next gather all suggestions that are not in that list
+        $suggesters = array();
+        $suggested = array();
+        foreach ($packages as $package) {
+            $packageName = $package['name'];
+            if ((!empty($filter) && !in_array($packageName, $filter)) || empty($package['suggest'])) {
```
Comment (samdark) [inline src/Composer/Command/SuggestsCommand.php]: `(!empty($filter)` could be replaced with `$filter !== array()`.

Comment (samdark) [inline src/Composer/Command/SuggestsCommand.php]: Same about `$package['suggest']`.

Comment (curry684) [inline src/Composer/Command/SuggestsCommand.php]: To be honest I prefer the more verbose clear syntax here, in my opinion `if not empty $filter` reads more self-explanatory than `if $property equals an empty array`. Performance-wise I don't expect a difference either, and both are only executed `count($installedPackage)` times.

Comment (samdark) [inline src/Composer/Command/SuggestsCommand.php]: I'm not insisting, just a suggestion.

**Comment by Seldaek** [comment]
Thanks, I just made -v enable only --by-package because the double listing is a bit over the top IMO, but overall it looks a lot better :)

**Comment by curry684** [comment]
Agreed, I was considering reverting it to no op even as the other options are obvious enough :)

**Comment by Seldaek** [comment]
I think it's good as short-hand + BC

**Pull request merged**